signer/xpi/cose.go

```diff
@@ -0,0 +1,339 @@
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/cose.manifest from signed zip: %v", err)
+	}
+	coseMsgBytes, err := readFileFromZIP(signedFile, "META-INF/cose.sig")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/cose.sig from signed zip: %v", err)
+	}
+	pkcs7Manifest, err := readFileFromZIP(signedFile, "META-INF/manifest.mf")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/manifest.mf from signed zip: %v", err)
+	}
+
+	var coseFileNames = [][]byte{
+		[]byte("Name: META-INF/cose.sig"),
+		[]byte("Name: META-INF/cose.manifest"),
+	}
+	for _, coseFileName := range coseFileNames {
+		if !bytes.Contains(pkcs7Manifest, coseFileName) {
+			return fmt.Errorf("xpi: pkcs7 manifest does not contain the line: %s", coseFileName)
+		}
+
+		if bytes.Contains(coseManifest, coseFileName) {
+			return fmt.Errorf("xpi: cose manifest contains the line: %s", coseFileName)
+		}
+	}
+
+	xpiSig, unmarshalErr := Unmarshal(base64.StdEncoding.EncodeToString(coseMsgBytes), nil)
+	if unmarshalErr != nil {
+		return errors.Wrap(unmarshalErr, "xpi: error unmarshaling cose.sig")
+	}
+	if xpiSig != nil && xpiSig.signMessage != nil && len(xpiSig.signMessage.Signatures) != len(signOptions.COSEAlgorithms) {
+		return fmt.Errorf("xpi: cose.sig contains %d signatures, but expected %d", len(xpiSig.signMessage.Signatures), len(signOptions.COSEAlgorithms))
+	}
+
+	intermediateCerts, eeCerts, err := isValidCOSEMessage(xpiSig.signMessage)
+	if err != nil {
+		return errors.Wrap(err, "xpi: cose.sig is not a valid COSE SignMessage")
+	}
+
+	// check that we can verify EE certs with the provided intermediates
+	intermediates := x509.NewCertPool()
+	for _, intermediateCert := range intermediateCerts {
+		intermediates.AddCert(intermediateCert)
+	}
+	cndigest := sha256.Sum256([]byte(signOptions.ID))
+	dnsName := fmt.Sprintf("%x.%x.addons.mozilla.org", cndigest[:16], cndigest[16:])
+
+	for i, eeCert := range eeCerts {
+		opts := x509.VerifyOptions{
+			DNSName:       dnsName,
+			Roots:         truststore,
+			Intermediates: intermediates,
+		}
+		if _, err := eeCert.Verify(opts); err != nil {
+			return fmt.Errorf("failed to verify EECert %d %s", i, err)
+		}
+	}
+	return nil
+}
+
+// issueCOSESignature returns a CBOR-marshalled COSE SignMessage
+// after generating EE certs and signatures for the COSE algorithms
+func (s *XPISigner) issueCOSESignature(cn string, manifest []byte, algs []*cose.Algorithm) (coseSig []byte, err error) {
+	if s == nil {
+		return nil, errors.New("Cannot issue COSE Signature from nil XPISigner")
+	}
+	if s.issuerCert == nil {
+		return nil, errors.New("Cannot issue COSE Signature when XPISigner.issuerCert is nil")
+	}
+
+	var (
+		coseSigners []cose.Signer
+		tmp         = cose.NewSignMessage()
+		msg         = &tmp
+	)
+	msg.Payload = manifest
+
+	// Add list of DER encoded intermediate certificates as message key id
+	msg.Headers.Protected["kid"] = [][]byte{s.issuerCert.Raw[:]}
```

Comment:
you already verify that `issuerCert != nil` above, so this is more of a nit, but maybe check that `len(s.issuerCert.Raw) > X` where X is a minimal value for a cert?